sql: clarify usage of the PK of virtual tables in some cases

yuzefovich opened the pull request:
The execution infrastructure for the virtual tables has a few limitations, one of them being not able to utilize the virtual indexes for range scans (i.e. only point lookups are supported). We do so transparently by handling the prefix of point lookups, and once we find the first range span in the index constraint, we fall back to the "full scan" of the PK. Previously, the EXPLAIN output would always show that we're using the constrained virtual index scan which can be misleading. This commit adjusts the EXPLAIN output to always show usage of the PK and to omit the spans altogether.

I considered applying the adjustment earlier, in the execbuilder, so that the virtual index wouldn't be chosen in the first place (which would effectively push the knowledge of the limitation of the execution higher), but I ran into a couple of difficulties:
- we would need to also teach the explain and plan-gist factories about the adjustment (because they capture their arguments before calling the wrapped factory)
- the execution infra currently assumes that whenever a constrained scan is used, we must be using a virtual secondary index. Making adjustments to this seems tricky and somewhat risky.

Fixes: #108317.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/126366)
<!-- Reviewable:end -->

yuzefovich committed:
sql: clarify usage of the PK of virtual tables in some cases

The execution infrastructure for the virtual tables has a few
limitations, one of them being not able to utilize the virtual indexes
for range scans (i.e. only point lookups are supported). We do so
transparently by handling the prefix of point lookups, and once we find
the first range span in the index constraint, we fall back to the "full
scan" of the PK. Previously, the EXPLAIN output would always show that
we're using the constrained virtual index scan which can be misleading.
This commit adjusts the EXPLAIN output to always show usage of the PK
and to omit the spans altogether.

I considered applying the adjustment earlier, in the execbuilder, so
that the virtual index wouldn't be chosen in the first place (which
would effectively push the knowledge of the limitation of the execution
higher), but I ran into a couple of difficulties:
- we would need to also teach the explain and plan-gist factories about
the adjustment (because they capture their arguments before calling
the wrapped factory)
- the execution infra currently assumes that whenever a constrained scan
is used, we must be using a virtual secondary index. Making adjustments
to this seems tricky and somewhat risky.

Release note: None

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice! Just one suggestion which you can take or leave.

> Reviewed 8 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126366#-:-O0jTvRe5SJYsbk3QIQ8:br84waw)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @yuzefovich)

___
*[`pkg/sql/opt/exec/explain/emit.go` line 648 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126366#-O0jT1P81svoaay5MTHf:-O0jT1P81svoaay5MTHg:b-p50xh3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c950ab6ab72e1e0a95cc782ac31ac3e38abee5af/pkg/sql/opt/exec/explain/emit.go#L648)):*
> ```Go
> 		}
> 		if a.Table.IsVirtualTable() && MaybeAdjustVirtualIndexScan != nil {
> 			a.Index, a.Params = MaybeAdjustVirtualIndexScan(ctx, evalCtx, a.Index, a.Params)
> ```

suggestion: Since the virtual table scan is also able to perform the filtering itself, internally, it might be nice to indicate that by adding something like `ob.Attr("virtual table filter", ...)`. (I'm not sure what it would take to build an expression that matches the virtual table filter, though.)


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I wonder, though, if it wouldn't be best to avoid generating constrained scans using ranges for virtual tables in the first place? I guess in the current situation we can handle a case with both point and range lookups.

> Reviewed 8 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126366#-:-O0kLj1l0c4GWjghRxsl:b-dhle9s)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
I prototyped not generating virtual index scans in the optimizer when execution doesn't support them in #126855 - please take a look there to see whether we should proceed with that approach (which makes sense to me).

mgartner commented:
@yuzefovich can we merge this and close #108317?

yuzefovich reviewed (commented):
@michae2 @DrewKimball thoughts about merging this? #126855 turned out tricky to get right, and this is an improvement to the status quo, so I think we should proceed.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126366#-:-OAZiO4p46HWwyrAsESc:bt3q1dt)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/opt/exec/explain/emit.go` line 648 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126366#-O0jT1P81svoaay5MTHf:-OAZfoduEYWX9-zLqEfl:b4gg8ci) ([raw file](https://github.com/cockroachdb/cockroach/blob/c950ab6ab72e1e0a95cc782ac31ac3e38abee5af/pkg/sql/opt/exec/explain/emit.go#L648)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

suggestion: Since the virtual table scan is also able to perform the filtering itself, internally, it might be nice to indicate that by adding something like `ob.Attr("virtual table filter", ...)`. (I'm not sure what it would take to build an expression that matches the virtual table filter, though.)
</blockquote></details>

Hm, you're referring to the filtering we do in `addRowIfPassesFilter`, right? There, before emitting a row that came from the virtual table we verify that it satisfies the index constraint if we have one (if not, that row is skipped). We could include the spans from the index constraint into the EXPLAIN output, but I don't see an easy way construct a full expression, so we could easily do something like:
```
EXPLAIN SELECT * FROM crdb_internal.system_jobs WHERE id > 100;
...
            distribution: local
            vectorized: true
            ·
            • virtual table
              table: system_jobs@primary
              constraint: [/101 - ]
```
but this wouldn't mention which column we're constraining. I probably would find this output more confusing than without the `constraint: ...` line. WDYT?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
Yeah, let's get this merged.

> Reviewed 8 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126366#-:-OAZlVKED8PlbXAyquCx:buorl1r)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @michae2)



<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Ship it!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126366#-:-OAZpwa_4a_eLOdWmpb2:b-psg8hr)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich)

___
*[`pkg/sql/opt/exec/explain/emit.go` line 648 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126366#-O0jT1P81svoaay5MTHf:-OAZndat6XFeFZQGy7x9:b-7hhd3v) ([raw file](https://github.com/cockroachdb/cockroach/blob/c950ab6ab72e1e0a95cc782ac31ac3e38abee5af/pkg/sql/opt/exec/explain/emit.go#L648)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, you're referring to the filtering we do in `addRowIfPassesFilter`, right? There, before emitting a row that came from the virtual table we verify that it satisfies the index constraint if we have one (if not, that row is skipped). We could include the spans from the index constraint into the EXPLAIN output, but I don't see an easy way construct a full expression, so we could easily do something like:
```
EXPLAIN SELECT * FROM crdb_internal.system_jobs WHERE id > 100;
...
            distribution: local
            vectorized: true
            ·
            • virtual table
              table: system_jobs@primary
              constraint: [/101 - ]
```
but this wouldn't mention which column we're constraining. I probably would find this output more confusing than without the `constraint: ...` line. WDYT?
</blockquote></details>

I agree, anything that looks like a span would be confusing, since this isn't a constrained scan.

What about adding just the string const `virtual table filter` to the output if `MaybeAdjustVirtualIndexScan` changed a non-nil index constraint to nil?

(I'm just bikeshedding, you can leave this the way it is if you want.)


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126366#-:-OAZvZ722xFWaw-GbOUa:b-tlx7ko)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/opt/exec/explain/emit.go` line 648 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126366#-O0jT1P81svoaay5MTHf:-OAZvHbS1fu7ShwACdrM:b-ls2id4) ([raw file](https://github.com/cockroachdb/cockroach/blob/c950ab6ab72e1e0a95cc782ac31ac3e38abee5af/pkg/sql/opt/exec/explain/emit.go#L648)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I agree, anything that looks like a span would be confusing, since this isn't a constrained scan.

What about adding just the string const `virtual table filter` to the output if `MaybeAdjustVirtualIndexScan` changed a non-nil index constraint to nil?

(I'm just bikeshedding, you can leave this the way it is if you want.)
</blockquote></details>

I'm neutral on this, but I don't think it will hurt, so I added it. Done.


<!-- Sent from Reviewable.io -->
